pumpkin-world/src/chunk/anvil.rs

```diff
@@ -141,41 +186,300 @@ impl ChunkReader for AnvilChunkReader {
+        location_table[table_entry as usize..table_entry as usize + 4]
+            .as_mut()
+            .copy_from_slice(&[location_bytes, &size_bytes].concat());
+
+        // write new location and timestamp table
+
+        region_file.seek(SeekFrom::Start(0)).unwrap(); // TODO
+        if let Err(err) = region_file.write_all(&[location_table, timestamp_table].concat()) {
+            return Err(ChunkWritingError::IoError(err.kind()));
+        }
+        // dbg!(&location_table.iter().map(|v| v.to_string()).join(","));
+
+        region_file.seek(SeekFrom::Start(end_index)).unwrap(); // TODO
+
+        let mut header: BytesMut = BytesMut::with_capacity(5);
+        // total chunk size includes the byte representing the compression
+        // scheme, so +1.
+        header.put_u32(bytes.len() as u32 + 1);
+        // compression scheme
+        header.put_u8(compression as u8);
+        region_file
+            .write_all(&header)
+            .expect("Failed to write header");
+        region_file.write_all(&bytes).unwrap_or_else(|_| {
+            panic!(
+                "Region file r.-{},{}.mca got corrupted, sorry",
+                region.0, region.1
+            )
+        });
+
+        region_file
+            .write_all(&vec![0u8; 4096])
+            .expect("Failed to add padding");
+
+        Ok(())
+    }
+}
+
+impl AnvilChunkFormat {
+    pub fn to_bytes(&self, chunk_data: &ChunkData) -> Result<Vec<u8>, ChunkSerializingError> {
+        let mut sections = Vec::new();
+
+        for (i, blocks) in chunk_data.blocks.blocks.chunks(16 * 16 * 16).enumerate() {
+            // get unique blocks
+            let palette = HashMap::<u16, (&String, usize)>::from_iter(
+                blocks.iter().enumerate().map(|(i, v)| {
+                    (
+                        *v,
+                        (
+                            BLOCK_ID_TO_REGISTRY_ID
+                                .get(v)
+                                .expect("Tried saving a block which does not exist."),
+                            i,
+                        ),
+                    )
+                }),
+            );
+
+            let block_bit_size = {
+                let size = 64 - (palette.len() as i64 - 1).leading_zeros();
+                std::cmp::max(4, size)
+            } as usize;
+            let blocks_in_pack = 64 / block_bit_size;
+            let mut section_longs = Vec::new();
+
+            let mut current_pack_long = 0i64;
+
+            for block_pack in blocks.chunks(blocks_in_pack) {
+                for block in block_pack {
+                    let index = palette.get(block).expect("Just added all unique").1;
+                    current_pack_long = current_pack_long << block_bit_size | index as i64;
+                }
+                section_longs.push(current_pack_long);
+                current_pack_long = 0;
+            }
+
+            sections.push(ChunkSection {
+                y: i as i8,
```

Comment:
There can be -y heights, but `i` is a usize. Should there be a start_y passed in?